File: VB/Reporting-Provide-Authentication-to-Access-JSON-Data/ReportCustomization/XtraReprotWizardCustomization.vb
Imports DevExpress.DataAccess.Json
Imports DevExpress.DataAccess.Native.Json
Imports DevExpress.DataAccess.UI.Json
Imports DevExpress.DataAccess.UI.Native.Json
Imports DevExpress.DataAccess.UI.Wizard
Imports DevExpress.DataAccess.Wizard
Imports DevExpress.DataAccess.Wizard.Model
Imports DevExpress.DataAccess.Wizard.Presenters
Imports DevExpress.DataAccess.Wizard.Views
Imports DevExpress.XtraLayout
Imports DevExpress.XtraReports.Wizards
Imports System
Imports System.ComponentModel.Design

Namespace XtraReport_JsonDataSource_with_Authorization.ReportCustomization
	Public Interface IMyChooseJsonSourcePageView
		Inherits IChooseJsonSourcePageView

		Property UserName() As String
		Property Password() As String
	End Interface
	Public Class MyChooseJsonSourcePageView
		Inherits DevExpress.DataAccess.UI.Wizard.Views.ChooseJsonSourcePageView
		Implements IMyChooseJsonSourcePageView

		Private Property IChooseJsonSourcePageView_ConnectionType() As JsonConnectionType Implements IChooseJsonSourcePageView.ConnectionType
			Get
				Return MyBase.ConnectionType
			End Get
			Set
				MyBase.ConnectionType = value
			End Set
		End Property

		Private Property IChooseJsonSourcePageView_Json() As String Implements IChooseJsonSourcePageView.Json
			Get
				Return MyBase.Json
			End Get
			Set
				MyBase.Json = value
			End Set
		End Property

		Private Property IChooseJsonSourcePageView_Uri() As String Implements IChooseJsonSourcePageView.Uri
			Get
				Return MyBase.Uri
			End Get
			Set
				MyBase.Uri = value
			End Set
		End Property

		Private Property IChooseJsonSourcePageView_FilePath() As String Implements IChooseJsonSourcePageView.FilePath
			Get
				Return MyBase.FilePath
			End Get
			Set
				MyBase.FilePath = value
			End Set
		End Property

		Public Property UserName() As String Implements IMyChooseJsonSourcePageView.UserName
			Get
				Return userNameCtrl.Text
			End Get
			Set(ByVal value As String)
				userNameCtrl.Text = value
			End Set
		End Property
		Public Property Password() As String Implements IMyChooseJsonSourcePageView.Password
			Get
				Return PasswordCtrl.Text
			End Get
			Set(ByVal value As String)
				PasswordCtrl.Text = value
			End Set
		End Property

		Protected ReadOnly userNameLayoutItem As LayoutControlItem
		Protected ReadOnly passwordLayoutItem As LayoutControlItem
		Protected ReadOnly userNameCtrl As DevExpress.XtraEditors.TextEdit
		Protected ReadOnly PasswordCtrl As DevExpress.XtraEditors.TextEdit
		Public Sub New()
			MyBase.New()
			userNameLayoutItem = layoutControlContent.AddItem()
			userNameLayoutItem.Text = "User Name"
			userNameCtrl = New DevExpress.XtraEditors.TextEdit()
			userNameLayoutItem.Control = userNameCtrl

			passwordLayoutItem = layoutControlContent.AddItem()
			passwordLayoutItem.Text = "Password"
			PasswordCtrl = New DevExpress.XtraEditors.TextEdit()
			PasswordCtrl.Properties.PasswordChar = "*"c
			PasswordCtrl.Properties.UseSystemPasswordChar = True
			passwordLayoutItem.Control = PasswordCtrl
		End Sub

		Protected Overrides Sub TypeComboOnSelectedValueChanged(ByVal sender As Object, ByVal e As EventArgs)
			MyBase.TypeComboOnSelectedValueChanged(sender, e)
			If userNameLayoutItem Is Nothing OrElse passwordLayoutItem Is Nothing Then
				Return
			End If
			If ConnectionType <> JsonConnectionType.Uri Then
				userNameLayoutItem.Visibility = DevExpress.XtraLayout.Utils.LayoutVisibility.Never
				passwordLayoutItem.Visibility = DevExpress.XtraLayout.Utils.LayoutVisibility.Never
			Else
				userNameLayoutItem.Visibility = DevExpress.XtraLayout.Utils.LayoutVisibility.Always
				passwordLayoutItem.Visibility = DevExpress.XtraLayout.Utils.LayoutVisibility.Always
			End If
		End Sub
	End Class
	Public Class MyChooseJsonSourcePage(Of TModel As {Class, IJsonDataSourceModel})
		Inherits ChooseJsonSourcePage(Of TModel)

		Private Context As IWizardRunnerContext
		Public Sub New(ByVal view As IChooseJsonSourcePageView, ByVal context As IWizardRunnerContext)
			MyBase.New(view, context)
			Me.Context = context
		End Sub
		Public Overrides Sub Begin()
			MyBase.Begin()
			If Model.JsonSource IsNot Nothing Then
				Dim myJsonSource = TryCast(Model.JsonSource, MyUriJsonSource)
				If myJsonSource IsNot Nothing AndAlso TypeOf View Is IMyChooseJsonSourcePageView Then
					TryCast(View, IMyChooseJsonSourcePageView).Password = myJsonSource.Password
					TryCast(View, IMyChooseJsonSourcePageView).UserName = myJsonSource.UserName
				End If
			End If
		End Sub
		Public Overrides Function Validate(<System.Runtime.InteropServices.Out()> ByRef errorMessage As String) As Boolean
			If View.ConnectionType <> JsonConnectionType.Uri Then
				Return MyBase.Validate(errorMessage)
			Else
				errorMessage = Nothing
				Dim source As JsonDataSource = Nothing
				Try
					Return AsyncHelper.DoWithWaitForm(Sub(token)
						source = New JsonDataSource With {
							.JsonSource = New MyUriJsonSource() With {
								.Uri = New Uri(View.Uri),
								.UserName = (TryCast(View, IMyChooseJsonSourcePageView))?.UserName,
								.Password = (TryCast(View, IMyChooseJsonSourcePageView))?.Password
							}
						}
						token.ThrowIfCancellationRequested()
						source.Fill()
						Model.DataSchema = DevExpress.DataAccess.Native.Data.DataView.ConvertToDataView(source)
						Model.JsonSchema = source.Schema
					End Sub, Context.WaitFormActivator, Context.CreateExceptionHandler(ExceptionHandlerKind.Loading))
				Catch e As Exception
					errorMessage = e.Message
					Return False
				Finally
					source?.Dispose()
				End Try
			End If
		End Function
		Public Overrides Sub Commit()
			Model.JsonSource = New MyUriJsonSource() With {
				.Uri = New Uri(View.Uri),
				.UserName = (TryCast(View, IMyChooseJsonSourcePageView))?.UserName,
				.Password = (TryCast(View, IMyChooseJsonSourcePageView))?.Password
			}
			RemoveHandler View.Changed, AddressOf ViewOnChanged
		End Sub
	End Class
	Public Class MyWizardCustomizationService
		Implements IWizardCustomizationService, IJsonEditorsCustomizationService

		Public Sub CustomizeDataSourceWizard(ByVal tool As IWizardCustomization(Of XtraReportModel)) Implements IWizardCustomizationService.CustomizeDataSourceWizard
			tool.RegisterPage(Of ChooseJsonSourcePage(Of XtraReportModel), MyChooseJsonSourcePage(Of XtraReportModel))()
			tool.RegisterPageView(Of IChooseJsonSourcePageView, MyChooseJsonSourcePageView)()
		End Sub

		Public Sub CustomizeReportWizard(ByVal tool As IWizardCustomization(Of XtraReportModel)) Implements IWizardCustomizationService.CustomizeReportWizard
			tool.RegisterPage(Of ChooseJsonSourcePage(Of XtraReportModel), MyChooseJsonSourcePage(Of XtraReportModel))()
			tool.RegisterPageView(Of IChooseJsonSourcePageView, MyChooseJsonSourcePageView)()
		End Sub

		Public Sub CustomizeWizard(ByVal editor As JsonEditorId, ByVal tool As IWizardCustomization(Of JsonDataSourceModel)) Implements IJsonEditorsCustomizationService.CustomizeWizard
			tool.RegisterPage(Of ChooseJsonSourcePage(Of JsonDataSourceModel), MyChooseJsonSourcePage(Of JsonDataSourceModel))()
			tool.RegisterPageView(Of IChooseJsonSourcePageView, MyChooseJsonSourcePageView)()
		End Sub

		Public Function TryCreateDataSource(ByVal model As IDataSourceModel, <System.Runtime.InteropServices.Out()> ByRef dataSource As Object, <System.Runtime.InteropServices.Out()> ByRef dataMember As String) As Boolean Implements IWizardCustomizationService.TryCreateDataSource
			dataSource = Nothing
			dataMember = model?.DataMember
			Return False
		End Function

		Public Function TryCreateReport(ByVal designerHost As IDesignerHost, ByVal model As XtraReportModel, ByVal dataSource As Object, ByVal dataMember As String) As Boolean Implements IWizardCustomizationService.TryCreateReport
			Return False
		End Function
	End Class
End Namespace

File: CS/Reporting-Provide-Authentication-to-Access-JSON-Data/ReportCustomization/XtraReprotWizardCustomization.cs
using DevExpress.DataAccess.Json;
using DevExpress.DataAccess.Native.Json;
using DevExpress.DataAccess.UI.Json;
using DevExpress.DataAccess.UI.Native.Json;
using DevExpress.DataAccess.UI.Wizard;
using DevExpress.DataAccess.Wizard;
using DevExpress.DataAccess.Wizard.Model;
using DevExpress.DataAccess.Wizard.Presenters;
using DevExpress.DataAccess.Wizard.Views;
using DevExpress.XtraLayout;
using DevExpress.XtraReports.Wizards;
using System;
using System.ComponentModel.Design;

namespace XtraReport_JsonDataSource_with_Authorization.ReportCustomization
{
    public interface IMyChooseJsonSourcePageView : IChooseJsonSourcePageView {
        string UserName { get; set; }
        string Password { get; set; }
    }
    public class MyChooseJsonSourcePageView : DevExpress.DataAccess.UI.Wizard.Views.ChooseJsonSourcePageView, IMyChooseJsonSourcePageView {
        public string UserName {
            get {
                return userNameCtrl.Text;
            }
            set {
                userNameCtrl.Text = value;
            }
        }
        public string Password {
            get {
                return PasswordCtrl.Text;
            }
            set {
                PasswordCtrl.Text = value;
            }
        }

        protected readonly LayoutControlItem userNameLayoutItem;
        protected readonly LayoutControlItem passwordLayoutItem;
        protected readonly DevExpress.XtraEditors.TextEdit userNameCtrl;
        protected readonly DevExpress.XtraEditors.TextEdit PasswordCtrl;
        public MyChooseJsonSourcePageView() : base() {
            userNameLayoutItem = layoutControlContent.AddItem();
            userNameLayoutItem.Text = "User Name";
            userNameCtrl = new DevExpress.XtraEditors.TextEdit();
            userNameLayoutItem.Control = userNameCtrl;

            passwordLayoutItem = layoutControlContent.AddItem();
            passwordLayoutItem.Text = "Password";
            PasswordCtrl = new DevExpress.XtraEditors.TextEdit();
            PasswordCtrl.Properties.PasswordChar = '*';
            PasswordCtrl.Properties.UseSystemPasswordChar = true;
            passwordLayoutItem.Control = PasswordCtrl;
        }

        protected override void TypeComboOnSelectedValueChanged(object sender, EventArgs e) {
            base.TypeComboOnSelectedValueChanged(sender, e);
            if(userNameLayoutItem == null || passwordLayoutItem == null) {
                return;
            }
            if(ConnectionType != JsonConnectionType.Uri) {
                userNameLayoutItem.Visibility = DevExpress.XtraLayout.Utils.LayoutVisibility.Never;
                passwordLayoutItem.Visibility = DevExpress.XtraLayout.Utils.LayoutVisibility.Never;
            } else {
                userNameLayoutItem.Visibility = DevExpress.XtraLayout.Utils.LayoutVisibility.Always;
                passwordLayoutItem.Visibility = DevExpress.XtraLayout.Utils.LayoutVisibility.Always;
            }
        }
    }
    public class MyChooseJsonSourcePage<TModel> : ChooseJsonSourcePage<TModel> where TModel : class, IJsonDataSourceModel {
        IWizardRunnerContext Context;
        public MyChooseJsonSourcePage(IChooseJsonSourcePageView view, IWizardRunnerContext context) : base(view, context) {
            this.Context = context;
        }
        public override void Begin() {
            base.Begin();
            if(Model.JsonSource != null) {
                var myJsonSource = Model.JsonSource as MyUriJsonSource;
                if(myJsonSource != null && View is IMyChooseJsonSourcePageView) {
                    (View as IMyChooseJsonSourcePageView).Password = myJsonSource.Password;
                    (View as IMyChooseJsonSourcePageView).UserName = myJsonSource.Username;
                }
            }
        }
        public override bool Validate(out string errorMessage) {
            if (View.ConnectionType != JsonConnectionType.Uri) {
                return base.Validate(out errorMessage);
            } else {
                errorMessage = null;
                JsonDataSource source = null;
                try {
                    return AsyncHelper.DoWithWaitForm(token => {
                        source = new JsonDataSource {
                            JsonSource = new MyUriJsonSource() {
                                Uri = new Uri(View.Uri),
                                Username = (View as IMyChooseJsonSourcePageView)?.UserName,
                                Password = (View as IMyChooseJsonSourcePageView)?.Password
                            }
                        };
                        token.ThrowIfCancellationRequested();
                        source.Fill();
                        Model.DataSchema = DevExpress.DataAccess.Native.Data.DataView.ConvertToDataView(source);
                        Model.JsonSchema = source.Schema;
                    }, Context.WaitFormActivator, Context.CreateExceptionHandler(ExceptionHandlerKind.Loading));
                } catch(Exception e) {
                    errorMessage = e.Message;
                    return false;
                } finally {
                    source?.Dispose();
                }
            }
        }
        public override void Commit() {
            Model.JsonSource = new MyUriJsonSource() {
                Uri = new Uri(View.Uri),
                Username = (View as IMyChooseJsonSourcePageView)?.UserName,
                Password = (View as IMyChooseJsonSourcePageView)?.Password
            };
            View.Changed -= ViewOnChanged;
        }
    }
    public class MyWizardCustomizationService : IWizardCustomizationService, IJsonEditorsCustomizationService {
        public void CustomizeDataSourceWizard(IWizardCustomization<XtraReportModel> tool) {
            tool.RegisterPage<ChooseJsonSourcePage<XtraReportModel>, MyChooseJsonSourcePage<XtraReportModel>>();
            tool.RegisterPageView<IChooseJsonSourcePageView, MyChooseJsonSourcePageView>();
        }

        public void CustomizeReportWizard(IWizardCustomization<XtraReportModel> tool) {
            tool.RegisterPage<ChooseJsonSourcePage<XtraReportModel>, MyChooseJsonSourcePage<XtraReportModel>>();
            tool.RegisterPageView<IChooseJsonSourcePageView, MyChooseJsonSourcePageView>();
        }

        public void CustomizeWizard(JsonEditorId editor, IWizardCustomization<JsonDataSourceModel> tool) {
            tool.RegisterPage<ChooseJsonSourcePage<JsonDataSourceModel>, MyChooseJsonSourcePage<JsonDataSourceModel>>();
            tool.RegisterPageView<IChooseJsonSourcePageView, MyChooseJsonSourcePageView>();
        }

        public bool TryCreateDataSource(IDataSourceModel model, out object dataSource, out string dataMember) {
            dataSource = null;
            dataMember = model?.DataMember;
            return false;
        }

        public bool TryCreateReport(IDesignerHost designerHost, XtraReportModel model, object dataSource, string dataMember) {
            return false;
        }
    }
}
